### PR TITLE
Searches

### PR DESCRIPTION
**Pull request by vojtik**

### CODE REVIEW

File: src/Repository/SearchInputRepository.php [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+     * @param array $params
+     *
+     * @return array
+     */
+    public function getSearchQuery($params)
+    {
+        $data['query'] = [];
+        $data['query']['ands'] = [];
+
+        if (isset($params[self::INPUT_CONCEPTS])) {
```
Comment (mxnr) [inline src/Repository/SearchInputRepository.php]: It's better to use array_key_exists in such cases

File: src/Repository/SearchInputRepository.php [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+
+    /**
+     * Generates Input Concept search query and adds it to existing data
+     *
+     * @param $data
+     * @param Concept[] $concepts
+     *
+     * @return array $data
+     */
+    public function generateInputConceptsQuery($data, $concepts)
```
Comment (mxnr) [inline src/Repository/SearchInputRepository.php]: This method and `generateOutputConceptsQuery` looks the same, split the same logic into different method

File: src/Repository/SearchInputRepository.php [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+
+    /**
+     * Generates Metadata search query and adds it to existing data
+     *
+     * @param $data
+     * @param array $metadata
+     *
+     * @return array $data
+     */
+    public function generateMetadataQuery($data, $metadata)
```
Comment (mxnr) [inline src/Repository/SearchInputRepository.php]: And this one

File: src/Repository/SearchInputRepository.php [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+
+    /**
+     * Generates Image search query and adds it to existing data
+     *
+     * @param $data
+     * @param Input[] $inputs
+     *
+     * @return array $data
+     */
+    public function generateImagesQuery($data, $inputs)
```
Comment (mxnr) [inline src/Repository/SearchInputRepository.php]: And this:)

File: src/Repository/SearchInputRepository.php [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+
+    /**
+     * Generates Reverse Image search query and adds it to existing data
+     *
+     * @param $data
+     * @param Input[] $inputs
+     *
+     * @return array $data
+     */
+    public function generateReverseImagesQuery($data, $inputs)
```
Comment (mxnr) [inline src/Repository/SearchInputRepository.php]: Aaand here. I consider to refactor this part, probably it should be moved to another class.

File: tests/Clarifai/Repository/SearchInputRepositoryTest.php [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+            $this->searchInputRepository->searchByUserSuppliedConcepts([$concept])
+        );
+    }
+
+    public function testSearchByReverseImage()
+    {
+        $input1 = $this->getFullInputEntity()->setId('id1')->setImage('image');
+        $input2 = $this->getFullInputEntity()->setId('id2')->setImage('image');
+
+        $this->request->shouldReceive('request')
```
Comment (mxnr) [inline tests/Clarifai/Repository/SearchInputRepositoryTest.php]: Also here you have a lot of almost the same arrays, just with one or two different keys. Consider to refactor them and move to helper method.

**Comment by darrynten** [comment]
The comments from @mxnr look good to me, looking forward to the update :)

File: src/Repository/SearchInputRepository.php (hunk 1 above)
Comment (vojtik) [inline src/Repository/SearchInputRepository.php]: array_key_exists will return true if $params[self::INPUT_CONCEPTS] === null and as a result we'll have one empty $data['query']['ands'] in our query and may cause error. In fact we do not need this key empty - only with data inside.

File: src/Repository/SearchInputRepository.php (hunk 5 above)
Comment (vojtik) [inline src/Repository/SearchInputRepository.php]: I've noticed that they are almost the same, but still couldn't figure out how to refactor them. they are using different types of data and sometimes different types of nesting. So it is or 5 almost same methods or one method with more than 5 if(){}'s

File: src/Repository/SearchInputRepository.php [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+     * @return array $data
+     */
+    public function generateImagesQuery($data, $inputs)
+    {
+        foreach ($inputs as $input) {
+            $data[] = [
+                'input' => [
+                    'data' => [
+                        'image' => [
+                            "url" => $input->getImage(),
```
Comment (mxnr) [inline src/Repository/SearchInputRepository.php]: Please use single quote here

File: src/Repository/SearchInputRepository.php [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+     */
+    public function generateReverseImagesQuery($data, $inputs)
+    {
+        foreach ($inputs as $input) {
+            $data[] = [
+                'output' => [
+                    'input' => [
+                        'data' => [
+                            'image' => [
+                                "url" => $input->getImage(),
```
Comment (mxnr) [inline src/Repository/SearchInputRepository.php]: And here

File: src/Repository/SearchModelRepository.php [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+     *
+     * @return Model[] array
+     */
+    public function searchByNameAndType($name = null, $type = null)
+    {
+        $searchResult = $this->getRequest()->request(
+            'POST',
+            'models/searches',
+            [
+                "model_query" => [
```
Comment (mxnr) [inline src/Repository/SearchModelRepository.php]: And here

File: tests/Clarifai/Repository/SearchModelRepositoryTest.php [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        $model2 = $this->getModelEntity()->setId('id2');
+        $model2->setRawData($model2->generateRawData());
+
+        $this->request->shouldReceive('request')
+            ->once()
+            ->with(
+                'POST',
+                'models/searches',
+                [
+                    "model_query" => [
```
Comment (mxnr) [inline tests/Clarifai/Repository/SearchModelRepositoryTest.php]: And also here.
Thanks!

**Comment by mxnr** [comment]
LGTM👍

**Comment by darrynten** [comment]
This is looking great! 

I think it's just that last check on Codacy left.

**Comment by vojtik** [comment]
It seems that not all status checks have passed - so I cannot merge

**Comment by darrynten** [comment]
It seems it was a bug with the checking service, everything was good.

**Pull request merged**